source/loader/layers/sanitizer/asan/asan_interceptor.cpp

```diff
@@ -639,22 +639,26 @@ ur_result_t AsanInterceptor::eraseProgram(ur_program_handle_t Program) {
     return UR_RESULT_SUCCESS;
 }
 
-ur_result_t AsanInterceptor::insertKernel(ur_kernel_handle_t Kernel) {
-    std::scoped_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
-    if (m_KernelMap.find(Kernel) != m_KernelMap.end()) {
-        return UR_RESULT_SUCCESS;
+KernelInfo &AsanInterceptor::getOrCreateKernelInfo(ur_kernel_handle_t Kernel) {
+    {
+        std::shared_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
+        if (m_KernelMap.find(Kernel) != m_KernelMap.end()) {
+            return *m_KernelMap[Kernel].get();
+        }
     }
 
-    auto hProgram = GetProgram(Kernel);
-    auto PI = getAsanInterceptor()->getProgramInfo(hProgram);
+    // Create new KernelInfo
+    auto Program = GetProgram(Kernel);
+    auto PI = getProgramInfo(Program);
     bool IsInstrumented = PI->isKernelInstrumented(Kernel);
 
+    std::scoped_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
     m_KernelMap.emplace(Kernel,
-                        std::make_shared<KernelInfo>(Kernel, IsInstrumented));
-    return UR_RESULT_SUCCESS;
+                        std::make_unique<KernelInfo>(Kernel, IsInstrumented));
```

Comment:
Why bother using a `unique_ptr` if there is no ownership change? We can just use raw pointer here.

Reply:
We need to manually delete them if it's raw pointer. `unique_ptr` is more convenient here.

Reply:
Then we can just use the object created by the map. It should be deleted when being erased from then map.

Reply:
> Then we can just use the object created by the map. It should be deleted when being erased from then map.

I've tried this but encountered a difficult-to-solve compilation error.
Using `unqiue_ptr` is a simple solution here.

Reply:
Then I am fine with this as a temporary solution. We plan to refactor related code in near future anyway. We can discuss for any better solution in that time.